Allow per deployment override of mem, disk and services

ericbottard opened the pull request:
fixes spring-cloud/spring-cloud-dataflow-admin-cloudfoundry#7

For example, to bind the `redis2` service to the `log` module, in addition to the `redis` service that would be bound to all modules, one would do:

```
stream create foobaba --definition "time | log"

stream deploy foobaba --properties "module.log.cloudfoundry.services=redis2"
```

markpollack commented:
Sabby wrote "We do not want to expose the bind to every module launched/orchestrated by admin."  How to change to redis2 at a more global level?

ericbottard commented:
Not sure I understand what you mean. The original issue is exactly about the opposite of "How to change to redis2 at a more global level?", ie we want **per-module** bindings.

To bind some services (especially the one used for the binder) to **all** modules, there has always been the **admin** property `cloudfoundry.services`. This adds a **deployment** property named `cloudfoundry.services` (which is additive)

Hope this makes sense

mbogoevici commented:
LGTM, will merge after testing.

mbogoevici commented on src/main/java/org/springframework/cloud/dataflow/module/deployer/cloudfoundry/ApplicationModuleDeployer.java:
```diff
@@ -163,6 +164,23 @@ public void run() {
 	}
 
+	private int deduceMemory(Map<String, String> deploymentProperties) {
+		String override = deploymentProperties.get("cloudfoundry.memory");
```
Kind of a nitpick, but we may want to start using constants at this point, just to be able to get a more centralized view of the supported options.

ericbottard replied:
Absolutely, but I believe I have created a separate issue for that (will look it up). We also want to use units parsing, IMHO

mbogoevici replied:
OK, cool. Yes, units would be a good thing. Maybe we can just piggyback on the already supported format in the manifest.  https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#disk-quota

mbogoevici replied:
Actually, using the manifest format wouldn't work, since the API requires a numeric value.

mbogoevici commented on src/main/java/org/springframework/cloud/dataflow/module/deployer/cloudfoundry/CloudFoundryModuleDeployerProperties.java:
```diff
@@ -82,6 +82,16 @@
+	/**
+	 * The amount of disk space (MB) to allocate, if not overridden per-module.
+	 */
+	private int disk;
```
We can provide a default value here as well or at least pass a null if zero. Not sure what's the effect of passing the zero value downstream.

mbogoevici replied:
Seems like there's no adverse reaction to passing zero, but to avoid any future surprises, I think it'd be better to simply pass `null` if the value is not a positive integer.

ericbottard commented:
I addressed the review comment about disk, choosing to set the default at 1024 to retain compatibility with what we had before (stepping through the code, seems disk can be set to `null` while memory can't, so chose consistency).
Also added more validation to properties

mbogoevici commented:
Cool, agree with consistency. Also 1024 is the default that gets set when the disk is set to null, so we can live with that for now.

mbogoevici commented:
Rebased, squashed, merged w/https://github.com/spring-cloud/spring-cloud-dataflow-admin-cloudfoundry/commit/e5cc1402d7837c7dc5ba783235904dd61cf1204f
